web/api/v1/api.go

```diff
@@ -1576,11 +1577,25 @@ func (api *API) cleanTombstones(r *http.Request) apiFuncResult {
 	return apiFuncResult{nil, nil, nil, nil}
 }
 
-func (api *API) respond(w http.ResponseWriter, data interface{}, warnings storage.Warnings) {
+func (api *API) respond(w http.ResponseWriter, data interface{}, warnings notes.Warnings) {
 	statusMessage := statusSuccess
+	isEmptyResult := false
+	structData, ok := data.(*queryData)
+	if ok {
+		isEmptyResult = structData.Result.String() == ""
+	}
 	var warningStrings []string
+	warningStringsMap := make(map[string]struct{})
```

Comment:
I think we would still need deduplication as the exact same warning would still appear multiple times for the same metric. You're right that it's not so ideal to do it here. What do you think of my suggestions under `Possible future improvements` in the PR description?

Reply:
If there aren't that many duplicates, it might even be OK to leave them in. But given that there might be a _lot_ of duplicates, it's even a resource usage problem to drag those many warnings through the whole stack, only to deduplicate them here, in the last step, taking a lot of resources again. I would rather go for a deduplication on the fly from the beginning to avoid that. (Note that this code lands in the main prometheus codebase and will affect many many users, of which some might get bitten hard if their Prometheus server suddenly takes some percent more RAM. And I would like to avoid hiding the enhanced warnings behind yet another feature flag.)